eclair-core/src/main/scala/fr/acinq/eclair/router/SyncHandlers.scala

```diff
@@ -0,0 +1,203 @@
+      Kamon.runWithSpan(Kamon.spanBuilder("reply-channel-range").start(), finishSpan = true) {
+
+        @tailrec
+        def loop(ids: List[ShortChannelId], timestamps: List[ReplyChannelRangeTlv.Timestamps], checksums: List[ReplyChannelRangeTlv.Checksums], acc: List[ShortChannelIdAndFlag] = List.empty[ShortChannelIdAndFlag]): List[ShortChannelIdAndFlag] = {
+          ids match {
+            case Nil => acc.reverse
+            case head :: tail =>
+              val flag = computeFlag(d.channels)(head, timestamps.headOption, checksums.headOption, routerConf.requestNodeAnnouncements)
+              // 0 means nothing to query, just don't include it
+              val acc1 = if (flag != 0) ShortChannelIdAndFlag(head, flag) :: acc else acc
+              loop(tail, timestamps.drop(1), checksums.drop(1), acc1)
+          }
+        }
+
+        val timestamps_opt = r.timestamps_opt.map(_.timestamps).getOrElse(List.empty[ReplyChannelRangeTlv.Timestamps])
+        val checksums_opt = r.checksums_opt.map(_.checksums).getOrElse(List.empty[ReplyChannelRangeTlv.Checksums])
+
+        val shortChannelIdAndFlags = Kamon.runWithSpan(Kamon.spanBuilder("compute-flags").start(), finishSpan = true) {
+          loop(r.shortChannelIds.array, timestamps_opt, checksums_opt)
+        }
+
+        val (channelCount, updatesCount) = shortChannelIdAndFlags.foldLeft((0, 0)) {
+          case ((c, u), ShortChannelIdAndFlag(_, flag)) =>
+            val c1 = c + (if (QueryShortChannelIdsTlv.QueryFlagType.includeChannelAnnouncement(flag)) 1 else 0)
+            val u1 = u + (if (QueryShortChannelIdsTlv.QueryFlagType.includeUpdate1(flag)) 1 else 0) + (if (QueryShortChannelIdsTlv.QueryFlagType.includeUpdate2(flag)) 1 else 0)
+            (c1, u1)
+        }
+        log.info(s"received reply_channel_range with {} channels, we're missing {} channel announcements and {} updates, format={}", r.shortChannelIds.array.size, channelCount, updatesCount, r.shortChannelIds.encoding)
+
+        def buildQuery(chunk: List[ShortChannelIdAndFlag]): QueryShortChannelIds = {
+          // always encode empty lists as UNCOMPRESSED
+          val encoding = if (chunk.isEmpty) EncodingType.UNCOMPRESSED else r.shortChannelIds.encoding
+          QueryShortChannelIds(r.chainHash,
+            shortChannelIds = EncodedShortChannelIds(encoding, chunk.map(_.shortChannelId)),
+            if (r.timestamps_opt.isDefined || r.checksums_opt.isDefined)
+              TlvStream(QueryShortChannelIdsTlv.EncodedQueryFlags(encoding, chunk.map(_.flag)))
+            else
+              TlvStream.empty
+          )
+        }
+
+        // we update our sync data to this node (there may be multiple channel range responses and we can only query one set of ids at a time)
+        val replies = shortChannelIdAndFlags
+          .grouped(routerConf.channelQueryChunkSize)
+          .map(buildQuery)
+          .toList
+
+        val (sync1, replynow_opt) = addToSync(d.sync, origin.nodeId, replies)
```

Comment:
Shouldn't we also move `addToSync` inside this file?

Reply:
Done with dc79e8d.